### ISSUE TITLE
TimeoutException when executing changelevel command in PHP

### ISSUE THREAD

**Issue by losttime** [koraktor/steam-condenser#221]
Using PHP to control TF2. After authenticating successfully, I'm doing this:

```
$server->rconExec('changelevel ctf_well')
```

But really, it does it for any map I choose. The server accepts the command and changes the level, but SteamCondenser generates a TimeoutException.

From what I can tell, the server sends back one packet with an empty body as a response to the changelevel command. SteamCondenser seems to be waiting for at least 3 packets before it considers the response done. Because a second packet never comes, the connection times out.

SteamCondenser code `lib/steam/servers/SourceServer.php`:

``` php
    $response = array();
    do {
        try {
            # This works the first time through
            # TimeoutException the second time through
            $responsePacket = $this->rconSocket->getReply();

            if ($responsePacket instanceof RCONAuthResponse) {
                $this->rconAuthenticated = false;
                throw new RCONNoAuthException();
            }
        } catch (SocketException $e) {
            if (defined('SOCKET_ECONNRESET') &&
                $e->getCode() == SOCKET_ECONNRESET) {
                $this->rconAuthenticated = false;
                throw new RCONNoAuthException();
            } else {
                throw $e;
            }
        }

        $response[] = $responsePacket->getResponse();
    } while(sizeof($response) < 3 || strlen($responsePacket->getResponse()) > 0);
```

Not sure why the `while` condition expects a minimum of 3 responses before allowing the loop to stop, but I'm not well-versed in the RCON spec.

**Comment by losttime** [comment]
After reading more about [Multiple-Packet Responses](https://developer.valvesoftware.com/wiki/Source_RCON_Protocol#Multiple-packet_Responses), I understand that the packet with the empty body signifies the end of response packets that are meaningful to the command that was executed.

I'll start working on a way to drop out of the while loop if the first response packet has an empty body.

**Comment by koraktor** [comment]
This is already done in the second part of the `while` condition.

The problem here is probably identical to that mentioned in other issues. Please have a look at the other issues to see what has already been done on those problems.

**Comment by losttime** [comment]
I'm sorry. I'm just not seeing it. Because the 2 while conditions are evaluated with an 'or' (`||`), if the first condition is met, the while loop will continue. So until I have received 3 responses the loop will continue, even if the 1st or 2nd responses are empty. Is my reasoning correct?

I'm not sure why the `sizeof($response) < 3` is in place. If it were removed, the loop would function as desired and wouldn't throw erroneous timeout exceptions.

**Comment by koraktor** [comment]
You're right. My mistake, I've been in a hurry. :flushed:

> […] I'm not well-versed in the RCON spec.

The main problem here is that there is no spec. Most of this has been found out through trial and error.
Please have a look at issue #190 for example.

The current behavior has been proven to work reliably for many situations… except for map changes and server restarts where problems can occur (even depending on server configuration).

As I cannot rule out that server behavior has been changed in a recent update of TF2 (or other games) I'll try to reevaluate on this issue.

**Comment by losttime** [comment]
I certainly appreciate all that trial and error work you and your contributors have done to create SteamCondenser. It's been so very useful to have available.

**Comment by koraktor** [comment]
Just checked this with a CS:GO server. Works as expected – or at least it worked the same way during my last tests. I'll check TF2 next.

**Comment by koraktor** [comment]
It seems like TF2 is much less responsive during map changes. It takes several seconds before it will respond to RCON requests again.

This let me think about how RCON termination should work and I think I've come up with a more reliable and even faster way of doing requests. Please have a look at the commit above for the result.

**Comment by koraktor** [comment]
Closing this. The fix was released with the newest 1.3 releases and should solve this issue.
